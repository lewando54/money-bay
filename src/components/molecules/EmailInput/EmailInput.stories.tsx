/* eslint-disable @typescript-eslint/consistent-type-assertions */
import { StoryObj, Meta } from '@storybook/react-native'

import EmailInput from './EmailInput'

const meta = {
	component: EmailInput,
	argTypes: {
        // To write
	}
} satisfies Meta<typeof EmailInput>

export default meta
type Story = StoryObj<typeof meta>

export const Default: Story = {
	args: {
        // To write
	}
}
