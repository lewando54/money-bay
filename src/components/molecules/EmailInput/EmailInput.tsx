import React, { useState } from 'react'
import { MAIN_DARK } from '../../../styling/GlobalTheme.style'
import Input from '../../atoms/Input/Input'
import { Feather } from '@expo/vector-icons'
import { StyleProp, ViewStyle, TextStyle } from 'react-native'

interface IEmailInputProps{
    value?: string
    placeholder?: string
    containerStyle?: StyleProp<ViewStyle>
    textStyle?: StyleProp<TextStyle>
    onChange?: (newValue: string) => void
    testID?: string
}

export default function EmailInput({
    value,
    placeholder='Email',
    containerStyle,
    textStyle,
    onChange,
    testID
}: IEmailInputProps){
    const [ isValid, setIsValid ] = useState(false)

    const handleOnChange = (newValue: string) => {
        if(onChange){
            onChange(newValue)
        }

        // eslint-disable-next-line no-control-regex
        if(newValue.match(/(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9]))\.){3}(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9])|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])/)){
            setIsValid(() => true)
        }
        else{
            setIsValid(() => false)
        }
    }

    return (
        <Input 
            value={value} 
            placeholder={placeholder}
            textStyle={textStyle}
            containerStyle={containerStyle}
            testID={testID}
            onChange={handleOnChange} 
            icon={<Feather name="check" size={14} color={MAIN_DARK} />} 
            isIconVisible={isValid}
        />
    )
}
