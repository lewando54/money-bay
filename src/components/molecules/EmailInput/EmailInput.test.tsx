import React, { useState } from 'react'
import renderer from 'react-test-renderer'
import { fireEvent, render } from '@testing-library/react-native'
import EmailInput from './EmailInput'

interface IInputMockupProps {
  value?: string
  placeholder?: string
  onChange?: (newValue: string) => void
  testID?: string
}

function StatefulInput({
    value,
    placeholder='Type something',
    testID
}: IInputMockupProps){
    const [ textValue, setTextValue ] = useState(value)

    return (
        <EmailInput 
            value={textValue}
            placeholder={placeholder}
            testID={testID}
            onChange={(newValue) => setTextValue(() => newValue)}
        />
    )
}

describe('EmailInput', () => {
    it('should render correctly', () => {
        const tree = renderer.create(<EmailInput />)
        expect(tree).toMatchSnapshot()
    })

    it('should render icon with checkmark when email is correct', () => {
        const { getByTestId, rerender } = render(<StatefulInput testID='inputtest'/>)
        fireEvent.changeText(getByTestId('inputtest'), 'lewando54@gmail.com')
        rerender(<StatefulInput testID='inputtest'/>)
        expect(getByTestId('input-icon')).toBeTruthy()
    })

    it('should not render icon with checkmark when email is not correctly typed', () => {
        const { getByTestId, rerender, queryByTestId } = render(<StatefulInput testID='inputtest'/>)
        fireEvent.changeText(getByTestId('inputtest'), 'lewando54@gmail.com')
        rerender(<StatefulInput testID='inputtest'/>)
        fireEvent.changeText(getByTestId('inputtest'), 'lewando54')
        rerender(<StatefulInput testID='inputtest'/>)
        expect(queryByTestId('input-icon')).toBeFalsy()
    })
})
