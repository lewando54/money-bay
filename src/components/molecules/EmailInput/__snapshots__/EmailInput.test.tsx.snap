// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailInput should render correctly 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "white",
        "borderRadius": 10,
        "height": "auto",
        "padding": 10,
        "width": "100%",
      },
      {
        "elevation": 20,
        "shadowColor": "#16196670",
        "shadowOffset": {
          "height": 6,
          "width": 0,
        },
        "shadowOpacity": 0.08,
        "shadowRadius": 16,
      },
      [
        {
          "alignItems": "center",
          "borderColor": "white",
          "borderWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingHorizontal": 14,
        },
        undefined,
        false,
      ],
    ]
  }
  testID="universal-container"
>
  <TextInput
    keyboardType="default"
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="Email"
    placeholderTextColor="#868698"
    secureTextEntry={false}
    style={
      [
        {
          "color": "white",
          "fontFamily": "MulishRegular",
        },
        {
          "color": "#1B1D4D",
          "marginRight": 10,
          "maxWidth": "93%",
          "width": "100%",
        },
        undefined,
        {
          "maxWidth": "100%",
        },
      ]
    }
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    hitSlop={10}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "display": "none",
      }
    }
    testID="input-icon"
  >
    <Text />
  </View>
</View>
`;
